modules/update-manifest-application.adoc

```diff
@@ -0,0 +1,59 @@
+// Module included in the following assemblies:
+//
+// * updating/understanding_updates/how-updates-work.adoc
+
+:_content-type: CONCEPT
+[id="update-manifest-application_{context}"]
+= Understanding how manifests are applied during an update
+
+Some manifests supplied in a release image need to be applied in a certain order because of the dependencies between them.
+For example, the `CustomResourceDefinition` resource must be created before the matching custom resources.
+Additionally, there is a logical order in which the individual Cluster Operators need to be updated to minimize disruption in the cluster.
+The Cluster Version Operator implements this logical order through the concept of Runlevels.
+
+These dependencies are encoded in the filenames of the manifests in the release image:
+
+[source, terminal]
+----
+0000_<runlevel>_<component>_<manifest-name>.yaml
+----
+
+For example:
+
+[source, terminal]
+----
+0000_03_config-operator_01_proxy.crd.yaml
+----
+
+The CVO internally builds a dependency graph for the manifests, obeying the following rules:
+
+* During an update, manifests at a lower Runlevel will be applied before those at a higher Runlevel.
+
+* Within one Runlevel, manifests for different components may be applied in parallel.
+
+* Within one Runlevel, manifests for a single component will be applied in lexicographic order.
+
+The CVO then applies manifests following the generated dependency graph.
+
+For some resource types, the CVO monitors the resource after its manifest is applied, and only considers it to be successfully updated after the resource reaches a stable state.
+Achieving this state can take some time.
+This is especially true for Cluster Operators, which may perform their own update actions in the cluster after the CVO deploys their new versions.
```

Comment:
nit: this is cheating a bit, which may be fine.  The structure is:

*  A component, e.g. "etcd".
  * [Assorted CVO-deployed manifests for that component][1], including:
    * [An operator Deployment][2].
    * [An associated ClusterOperator][3].
    * A bunch of other stuff (namespaces, RBAC, ServiceMonitors, etc.), possibly spread across multiple runlevels.  [Etcd is currently split][1] between levels 20 for the core componet and 90 for monitoring resources.
  * Other in-cluster resources created by the operator, once it starts running.

When you say "Cluster Operators" here, it's not clear if you mean "the whole component", "the set of all CVO-deployed manifests for that component", or "just [the ClusterOperator manifest][3]".  In the context of updates, the timing is usually:


1. CVO updates a bunch of resources like [the operator Deployment][2].
2. The updated component operator pods roll out their new version, and eventually update their ClusterOperator `status` to claim completion.
3. The CVO is not touching [the component's ClusterOperator][3], but is blocking on that manifest, until it is [satisfied][4] with the status.

Maybe we could rephrase with:

> Achieving this state can take some time, especially for ClusterOperator resources, when the CVO is waiting for a component operator to update its component and update its ClusterOperator `status`.

or something.

Also unclear to me if we are using "Cluster Operator" or "ClusterOperator" to refer to the ClusterOperator resources.  The current text seems to use both forms, and while there's no need to be perfectly consistent, the inconsistency may be contributing to my confusion over the precise meaning of "Cluster Operator".

[1]: https://github.com/openshift/cluster-etcd-operator/tree/c6c625427d4db4b26217dc304c2944df585fbb19/manifests
[2]: https://github.com/openshift/cluster-etcd-operator/blob/c6c625427d4db4b26217dc304c2944df585fbb19/manifests/0000_20_etcd-operator_06_deployment.yaml
[3]: https://github.com/openshift/cluster-etcd-operator/blob/c6c625427d4db4b26217dc304c2944df585fbb19/manifests/0000_20_etcd-operator_07_clusteroperator.yaml
[4]: https://github.com/openshift/enhancements/blob/a0482b15dfb9338ebbd6e7c6d711302febf78b99/dev-guide/cluster-version-operator/user/reconciliation.md#clusteroperator

Reply:
I'll save both feedback items in this comment for a second iteration.

Reply:
s/may perform/might perform